### PR TITLE
Add dedicated IBD rules for Nakamoto

### LINKED ISSUE #5642: [Bug] restore IBD calculation removed by #5505
This _must_ be fixed before 3.1.0.0.3, because it prevents nodes from advertising their StackerDB replicas.

### PR DESCRIPTION
**Pull request by jcnelson**
This fixes #5642 by adding a dedicated IBD inference rule for Nakamoto.  The Stacks node is in IBD mode when either of the following conditions are true:

* The sortition height is less than the Bitcoin block height
* The highest available tenure (as determined by the Nakamoto block downloader) is higher than the ongoing Stacks tenure 

This (hopefully) fixes some edge cases we've seen on testnet whereby a node can erroneously believe it is not sync'ed when it really is.  This had impacted the affected node's ability to participate in StackerDB replication.  I intend to test this on naka3.sh, on testnet, and on a single mainnet signer.

Writing the code for that second criterion led to the discoveries of #5649 and #5650, since I had been using `rc_consensus_hash` as the ongoing Stacks tenure when in fact it was not.

EDIT: this also contains #5667, so let's merge that first.

### CODE REVIEW

**Review by obycode — DISMISSED** [review]
LGTM!

File: stackslib/src/net/download/nakamoto/download_state_machine.rs [hunk 1]
```diff
@@ -1553,6 +1564,24 @@ impl NakamotoDownloadStateMachine {
...
+            self.prev_wanted_tenures.as_ref().unwrap_or(&vec![]),
+            &self.available_tenures,
+        );
+        if let Some(ch) = t2 {
+            return Some(ch);
+        } else if let Some(ch) = t1 {
+            return Some(ch);
+        } else {
+            return None;
+        }
```
Comment (jbencin) [inline stackslib/src/net/download/nakamoto/download_state_machine.rs:1582]: Could be reduced to:

```suggestion
        t2.or(t1)
```

File: stackslib/src/net/p2p.rs [hunk 2]
```diff
@@ -5258,6 +5258,7 @@ impl PeerNetwork {
         );
         let mut network_result = NetworkResult::new(
             self.stacks_tip.block_id(),
+            self.stacks_tip.consensus_hash.clone(),
```
Comment (jbencin) [inline stackslib/src/net/p2p.rs:5261]: `ConsensusHash` implements `Copy` so Clippy might complain about this once we enable it in CI. Can be:

```suggestion
            self.stacks_tip.consensus_hash,
```

File: stackslib/src/net/download/nakamoto/download_state_machine.rs [hunk 3]
```diff
@@ -862,6 +865,14 @@ impl NakamotoDownloadStateMachine {
         self.tenure_download_schedule = schedule;
         self.tenure_block_ids = tenure_block_ids;
         self.available_tenures = available;
+
+        let highest_available_tenure = self.find_highest_available_tenure();
+        self.highest_available_tenure = highest_available_tenure;
```
Comment (jbencin) [inline stackslib/src/net/download/nakamoto/download_state_machine.rs:870]: No need for the `let` binding here:

```suggestion
        self.highest_available_tenure = self.find_highest_available_tenure();
```

**Review by jbencin — DISMISSED** [review]
Left some style comments but looks fine to me

File: stackslib/src/net/download/nakamoto/download_state_machine.rs (hunk 3 above)
Comment (jcnelson) [reply stackslib/src/net/download/nakamoto/download_state_machine.rs:870]: Fixed

File: stackslib/src/net/p2p.rs [hunk 4]
```diff
@@ -5267,6 +5268,10 @@ impl PeerNetwork {
             self.stacks_tip.height,
             self.chain_view.rc_consensus_hash.clone(),
             self.get_stacker_db_configs_owned(),
+            self.block_downloader_nakamoto
+                .as_ref()
+                .map(|dler| dler.highest_available_tenure.clone())
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply stackslib/src/net/p2p.rs:5273]: Fixed

File: stackslib/src/net/p2p.rs (hunk 2 above)
Comment (jcnelson) [reply stackslib/src/net/p2p.rs:5261]: Fixed

File: stackslib/src/net/download/nakamoto/download_state_machine.rs (hunk 1 above)
Comment (jcnelson) [reply stackslib/src/net/download/nakamoto/download_state_machine.rs:1582]: Fixed

**Review by jferrant — DISMISSED** [review]

**Review by obycode — COMMENTED** [review]
Just some minor comments for now, but there are conflicts and failing tests.

File: testnet/stacks-node/src/nakamoto_node/miner.rs [hunk 5]
```diff
@@ -277,15 +275,15 @@ impl BlockMinerThread {
 
     #[cfg(test)]
     fn fault_injection_block_announce_stall(new_block: &NakamotoBlock) {
-        if TEST_BLOCK_ANNOUNCE_STALL.get() {
+        if relay::fault_injection::stacks_announce_is_blocked() {
```
Comment (kantai) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:278]: These seem like they could be independently controlled... `TEST_BLOCK_ANNOUNCE_STALL` is used in the integration tests to prevent the miner themselves from announcing a new block (this is used to test that the signer set can announce blocks by themselves) -- but if the relayer stall is also active, it seems like it prevents the chains coordinator thread from waking up?

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 6]
```diff
@@ -953,6 +952,8 @@ fn forked_tenure_testing(
         |config| {
             // make the duration long enough that the reorg attempt will definitely be accepted
             config.first_proposal_burn_block_timing = proposal_limit;
+            config.tenure_last_block_proposal_timeout = Duration::from_secs(0);
+
```
Comment (kantai) [inline testnet/stacks-node/src/tests/signer/v0.rs:956]: How is the `forked_tenure_testing` test related to this changeset? Why did this test need to be altered, and can you describe the intended behavior changes to the test?

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 7]
```diff
@@ -400,6 +400,67 @@ impl RelayerThread {
...
+            .bitcoin_controller
+            .get_headers_height()
+            .saturating_sub(1);
+
+        // are we still processing sortitions?
+        let burnchain_ibd = cur_sn.block_height != headers_height;
+
+        // if the highest available tenure is known, then is it the same as the ongoing stacks
+        // tenure?  If so, then we're not IBD. If not, then we're IBD.
+        // If it is not known, then we're not in IBD.
```
Comment (kantai) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:419]: Doesn't this mean that a running node operating at chain tip would switch into `IBD = true` at every tenure boundary?

**Review by kantai — COMMENTED** [review]
This changeset appears to have broken many of the integration tests.

This should also have some kind of unit test (or tight assertions in an integration test) about what the expected value of IBD should be given different values of the `NetworkResult`: its not clear to me what the intended behavior around the arrival of a new tenure is, and then what the downstream impact on miner commitments would end up being.

I also have some questions about the necessity of this -- it seems like #5735 actually resolves the testnet genesis sync issues (and the mainnet ones as well), so what is this PR solving? Is it speeding up genesis sync? Restarts? I can't tell from looking at this PR.

File: testnet/stacks-node/src/nakamoto_node/miner.rs (hunk 5 above)
Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/miner.rs:278]: At the time this PR was written, this was an attempt to make some CI tests less flaky.  It appears that not only is this not necessary anymore, but also my code here has led to CI breakage.

File: testnet/stacks-node/src/tests/signer/v0.rs (hunk 6 above)
Comment (jcnelson) [reply testnet/stacks-node/src/tests/signer/v0.rs:957]: I've reverted it -- it was part of me trying to un-flake this test from a while ago.

**Comment by jcnelson** [comment]
The system seems just fine without this patch.  Going to close.
